torch/distributed/_tools/mem_tracker.py

```diff
@@ -0,0 +1,920 @@
+    ) -> Tuple[Self, weakref.ref]:
+        """
+        Creates a new ``_WeakRefInfo`` instance and a weak reference to a ``torch.UntypedStorage`` object,
+        optionally attaching a callback to the weak reference.
+
+        Args:
+            st (torch.UntypedStorage): The storage object for which to create the weak reference info.
+            device (torch.device): The device associated with the storage object.
+            reftype (_RefType): The type of reference, used to categorize the storage.
+            callback (Optional[Callable[[Self, weakref.ref]]]): A callback function that is called when
+                the storage object is about to be finalized (garbage collected). The callback function
+                should accept two arguments: the ``_WeakRefInfo`` instance and the weak reference to the storage.
+        Returns:
+            Tuple[Self, weakref.ref]: A tuple containing the newly created ``_WeakRefInfo`` instance and the
+            weak reference to the storage object. The weak reference may have an attached callback if provided.
+        """
+
+        winfo = cls(st.size(), st.element_size(), device, reftype)
+        w_st = weakref.ref(st, partial(callback, winfo) if callback else None)
+        return winfo, w_st
+
+
+def _get_mem_divisor(units: str) -> int:
+    return {"B": 1, "KB": 2**10, "MB": 2**20, "GB": 2**30}.get(units, 1)
+
+
+def _rounding_fn(value: int, divisor: int, precision: int) -> Union[float, int]:
+    return value if divisor == 1 else round(value / divisor, precision)
+
+
+def _print_snapshot(snapshot: Dict[torch.device, Dict[str, int]], units: str) -> None:
+    if len(snapshot) == 0:
+        print("No memory tracked.")
+        return
+    divisor = _get_mem_divisor(units)
+    for dev, dev_snap in snapshot.items():
+        if _rounding_fn(dev_snap[_TOTAL_KEY], divisor, 2) <= 0:
+            continue
+        print(
+            f"Device: {dev}",
+            *(
+                f"\t{k}: {_rounding_fn(v, divisor, 2)} {units}"
+                for k, v in dev_snap.items()
+            ),
+            sep="\n",
+        )
+
+
+def _print_snapshot_tabular(
+    snapshot: Dict[torch.device, Dict[str, int]], units: str
+) -> None:
+    if len(snapshot) == 0:
+        print("No memory tracked.")
+        return
+    try:
+        from tabulate import tabulate
+
+        divisor = _get_mem_divisor(units)
+        table_data = []
+        key_list = list(next(iter(snapshot.values())).keys())
+        headers = ["Device"] + [f"{key}" for key in key_list]
+
+        for dev, dev_snap in snapshot.items():
+            if _rounding_fn(dev_snap[_TOTAL_KEY], divisor, 2) <= 0:
+                continue
+            row = [str(dev)]
+            row.extend(
+                f"{_rounding_fn(v, divisor, 2)} {units}" for v in dev_snap.values()
+            )
+            table_data.append(row)
+
+        print(tabulate(table_data, headers=headers, tablefmt="rst"))
+    except ImportError as err:
```

Comment:
nit: Would it be better to only enclose the `from tabulate import tabulate` in `try: except:`? (this is super minor though)